tests/test_all_integration.py

```diff
@@ -0,0 +1,316 @@
+    assert tx_receipt
+    # (1) Gateway_1 listens for the event
+    blockchain_gateway._listen(blockchain_gateway._handle_new_session_creation, 
+        blockchain_gateway._filter_new_session)
+    # (2) Gateway_2 listens for the event
+    blockchain_gateway_2._listen(blockchain_gateway_2._handle_new_session_creation, 
+        blockchain_gateway_2._filter_new_session)
+    # (3) Scheduler_1 and Scheduler_2 runs the following jobs:
+        # (3a.1) JOB_INIT
+        # (3a.2) JOB_SPLIT
+        # (3b) JOB_TRAIN
+        # (3c) JOB_COMM
+    scheduler.start_cron(period_in_mins = 0.01)
+    scheduler_2.start_cron(period_in_mins=0.01)
+    timeout = time.time() + 25
+    while time.time() < timeout and (len(scheduler.processed) != 4\
+        or len(scheduler_2.processed) != 4):
+        time.sleep(5)
+    assert len(scheduler.processed) == 4, "Jobs failed/not completed in time!"
+    assert len(scheduler_2.processed) == 4, "Jobs failed/not completed in time!"
+    # (4) Gateway_1 listens for the new weights 
+    blockchain_gateway._listen(blockchain_gateway._handle_new_session_info,
+        blockchain_gateway._filter_new_session_info)
+    # (6) Gateway_2 listens for the new weights
+    blockchain_gateway_2._listen(blockchain_gateway_2._handle_new_session_info,
+        blockchain_gateway_2._filter_new_session_info)
+    # (8) Scheduler_1 and Scheduler_2 runs the following jobs:
+        # (6a) JOB_AVG
+        # (6b) JOB_AVG
+        # (6a) JOB_TRAIN
+        # (6a) JOB_COMM
+    timeout = time.time() + 20
+    while time.time() < timeout and (len(scheduler.processed) != 8\
+        or len(scheduler_2.processed) != 8):
+        time.sleep(4)
+    assert len(scheduler.processed) == 8, \
+        "Jobs {} failed/not completed in time!".format([
+            result.job.job_type for result in scheduler.processed])
+    assert len(scheduler_2.processed) == 8, "Jobs failed/not completed in time!"
+    # (4) Gateway_1 listens for the new weights 
+    blockchain_gateway._listen(blockchain_gateway._handle_new_session_info,
+        blockchain_gateway._filter_new_session_info)
+    # (6) Gateway_2 listens for the new weights
+    blockchain_gateway_2._listen(blockchain_gateway_2._handle_new_session_info,
+        blockchain_gateway_2._filter_new_session_info)
+    # (13) Optimizer tells Communication Manager to schedule JOB_AVG
+    # (14) Scheduler_1 and Scheduler_2 runs the following jobs:
+        # (9a) JOB_AVG
+        # (9a) JOB_AVG
+    timeout = time.time() + 10
+    while time.time() < timeout and (len(scheduler.processed) != 10\
+        or len(scheduler_2.processed) != 10):
+        time.sleep(2)
+    scheduler.stop_cron()
+    scheduler_2.stop_cron()
+    assert len(scheduler.processed) == 10, "Jobs failed/not completed in time!"
+    assert len(scheduler_2.processed) == 10, "Jobs failed/not completed in time!"
+    # (10) Optimizer terminates
+    assert communication_manager.optimizer is None, "Should have terminated!"
```

Comment:
You can just do `not communication_manager.optimizer1

Reply:
Done